source/Orders.cpp

```diff
@@ -0,0 +1,315 @@
+}
+
+
+
+bool Orders::HasMine() const
+{
+	return activeOrders.test(OrderType::MINE);
+}
+
+
+
+bool Orders::HasHarvest() const
+{
+	return activeOrders.test(OrderType::HARVEST);
+}
+
+
+
+bool Orders::IsEmpty() const
+{
+	return activeOrders.none();
+}
+
+
+
+void Orders::SetTargetShip(shared_ptr<Ship> ship)
+{
+	targetShip = ship;
+}
+
+
+
+void Orders::SetTargetAsteroid(shared_ptr<Minable> asteroid)
+{
+	targetAsteroid = asteroid;
+}
+
+
+
+void Orders::SetTargetPoint(const Point &point)
+{
+	targetPoint = point;
+}
+
+
+
+shared_ptr<Ship> Orders::GetTargetShip() const
+{
+	return targetShip.lock();
+}
+
+
+
+void Orders::SetTargetSystem(const System *system)
+{
+	targetSystem = system;
+}
+
+
+
+shared_ptr<Minable> Orders::GetTargetAsteroid() const
+{
+	return targetAsteroid.lock();
+}
+
+
+
+const Point &Orders::GetTargetPoint() const
+{
+	return targetPoint;
+}
+
+
+
+const System *Orders::GetTargetSystem() const
+{
+	return targetSystem;
+}
+
+
+
+void Orders::UpdateOrder(const Ship *orderedShip, const System *flagshipSystem)
+{
+	if(HasMine() && orderedShip->Cargo().Free() && targetAsteroid.expired())
+		SetHarvest();
+	else if((activeOrders & REQUIRES_TARGET).any())
+	{
+		shared_ptr<Ship> ship = GetTargetShip();
+		shared_ptr<Minable> asteroid = GetTargetAsteroid();
+		// Check if the target ship itself is targetable, or if it is one of your ships that you targeted.
+		bool invalidTarget = !ship
+				|| (!ship->IsTargetable() && orderedShip->GetGovernment() != ship->GetGovernment())
+				|| (ship->IsDisabled() && HasAttack());
+		// Alternatively, if an asteroid is targeted, then it is not an invalid target.
+		invalidTarget &= !asteroid;
+
+		// Check if the target ship is in a system where we can target.
+		// This check only checks for undocked ships (that have a current system).
+		const System *orderSystem = orderedShip->GetSystem();
+		const System *shipSystem = ship ? ship->GetSystem() : nullptr;
+		bool targetOutOfReach = !ship || (orderSystem && shipSystem != orderSystem
+				&& shipSystem != flagshipSystem);
+		// Asteroids are never out of reach since they're in the same system as the flagship.
+		targetOutOfReach &= !asteroid;
+
+		// Cancel any orders that required a target.
+		if(invalidTarget || targetOutOfReach)
+			activeOrders &= ~REQUIRES_TARGET;
+	}
+}
+
+
+
+void Orders::MergeOrders(const Orders &other, bool &hasMismatch, bool &alreadyHarvesting, int &orderOperation)
+{
+	// HOLD_ACTIVE cannot be given as a manual order, but we make sure here
+	// that any HOLD_ACTIVE order also matches when a HOLD_POSITION
+	// command is given.
+	if(HasHoldActive())
+		ApplyOrder(HOLD_POSITION);
+
+	// For each bit of the other order that is flipped, apply the flip to this order.
+	for(size_t i = 0; i < other.activeOrders.size(); ++i)
+		if(other.activeOrders.test(i))
+		{
+			hasMismatch |= !activeOrders.test(i);
+			orderOperation = ApplyOrder(static_cast<OrderType>(i), orderOperation);
+		}
```

Comment:
New bug: if we focus fire on target A and then change to focus fire on target B, we shouldn't flip the attack bit off. The attack bit should remain on since we're still attacking, just attacking a separate target.